cmd/fleet/cron.go

```diff
@@ -19,105 +20,169 @@ import (
 	"github.com/go-kit/kit/log/level"
 )
 
-func cronDB(ctx context.Context, ds fleet.Datastore, logger kitlog.Logger, identifier string, license *fleet.LicenseInfo) {
-	logger = kitlog.With(logger, "cron", lockKeyLeader)
+func startWebhooksSchedule(
+	ctx context.Context, instanceID string, ds fleet.Datastore, appConfig *fleet.AppConfig, failingPoliciesSet fleet.FailingPolicySet, logger kitlog.Logger,
+) {
+	schedule.New(
+		ctx, "webhooks", instanceID, appConfig.WebhookSettings.Interval.ValueOr(24*time.Hour), ds,
+		schedule.WithLogger(kitlog.With(logger, "cron", "webhooks")),
+		schedule.WithConfigReloadInterval(5*time.Minute, func(ctx context.Context) (time.Duration, error) {
+			appConfig, err := ds.AppConfig(ctx)
+			if err != nil {
+				return 0, err
+			}
+			newInterval := appConfig.WebhookSettings.Interval.ValueOr(24 * time.Hour)
+			return newInterval, nil
+		}),
+		schedule.WithJob(
+			"maybe_trigger_host_status",
+			func(ctx context.Context) error {
+				return webhooks.TriggerHostStatusWebhook(
+					ctx, ds, kitlog.With(logger, "webhook", "host_status"), appConfig,
+				)
+			},
+		),
+		schedule.WithJob(
+			"maybe_trigger_failing_policies",
+			func(ctx context.Context) error {
+				return webhooks.TriggerFailingPoliciesWebhook(
+					ctx, ds, kitlog.With(logger, "webhook", "failing_policies"), appConfig, failingPoliciesSet, time.Now(),
+				)
+			},
+		),
+	).Start()
+}
 
-	ticker := time.NewTicker(10 * time.Second)
-	for {
-		level.Debug(logger).Log("waiting", "on ticker")
-		select {
-		case <-ticker.C:
-			level.Debug(logger).Log("waiting", "done")
-			ticker.Reset(1 * time.Hour)
-		case <-ctx.Done():
-			level.Debug(logger).Log("exit", "done with cron.")
-			return
-		}
+func startCleanupsAndAggregationSchedule(
+	ctx context.Context, instanceID string, ds fleet.Datastore, logger kitlog.Logger,
+) {
+	schedule.New(
+		ctx, "cleanups_then_aggregation", instanceID, 1*time.Hour, ds,
+		// Using leader for the lock to be backwards compatilibity with old deployments.
+		schedule.WithAltLockID("leader"),
```

Comment:
Do we need to use `WithAltLockID` here? what would happen if we just used the new lock name `cleanups_then_aggregation`?

Reply:
Let's discuss, I could alternatively use the name and have a small migration to remove the old lock from `locks` table.

Reply:
My only worry is that with how currently migrations work we don't have any assurances (there's no feature that allows blue/green deployment). A new Fleet server could be migrating and an old Fleet server could be running a cleanup with the old `leader` lock name at the same time (or be about to run cleanup with such lock name)...

Reply:
From the [docs](https://fleetdm.com/docs/deploying/upgrading-fleet#prepare-the-database)

> Database migrations in Fleet are intended to be run while the server is offline

If someone does perform migrations while fleet is running, this is not the only code that would break. Also, they would need to keep the old version of fleet running after performing migrations, which is also problematic.

Reply:
Agree, just didn't want to add (even) more risk to the release.

Hopefully we'll prioritize safe-migration in the near future 🤞